### PR TITLE
[OCPBUGS-25341]: perform operator apiService certificate validity checks directly

### PR DESCRIPTION
**Pull request by ankitathomas**
Cert updates can occasionally fail silently, updating only the timestamps on the CSV without any changes to the underlying cert secret. This PR uses the cert expiry times directly to retry the refresh.

### CODE REVIEW

File: test/e2e/webhook_e2e_test.go [hunk 1]
```diff
@@ -1169,3 +1181,70 @@ func newV1CRD(plural string) apiextensionsv1.CustomResourceDefinition {
...
+	Expect(err).Should(BeNil())
+
+	caDetails := &x509.Certificate{
+		SerialNumber: caSerial,
+		Subject: pkix.Name{
+			CommonName:   fmt.Sprintf("olm-selfsigned-%x", caSerial),
+			Organization: []string{install.Organization},
+		},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now(),
```
Comment (tmshort) [inline test/e2e/webhook_e2e_test.go:1196]: Wondering if you really want these times in the past...

File: pkg/controller/operators/olm/apiservices.go [hunk 2]
```diff
@@ -93,17 +93,12 @@ func (a *Operator) checkAPIServiceResources(csv *v1alpha1.ClusterServiceVersion,
...
 		if err != nil {
 			logger.Warnf("could not convert APIService CA bundle to x509 cert")
 			errs = append(errs, err)
 			continue
 		}
-		if !certs.Active(ca) {
-			logger.Warnf("CA cert not active")
-			errs = append(errs, fmt.Errorf("found the CA cert is not active"))
-			continue
-		}
```
Comment (tmshort) [inline pkg/controller/operators/olm/apiservices.go:106]: We're no longer checking the certs (here and below)?

Comment (ankitathomas) [reply pkg/controller/operators/olm/apiservices.go:106]: I've combined the cert expiry checks with [freshness](https://github.com/operator-framework/operator-lifecycle-manager/pull/3217/files#diff-bc7b779225ba70ff0e1f63ff55cb2dd3d71b5ae4f96c158947266da9f8fde60dR274) so we don't have a state needing manual cert refreshes for already expired certs

File: pkg/controller/install/certresources.go [hunk 3]
```diff
@@ -223,15 +230,74 @@ func (i *StrategyDeploymentInstaller) CertsRotated() bool {
...
+			return true
+		}
+	}
 	return false
 }
 
+func (i *StrategyDeploymentInstaller) ShouldRotateCerts(s Strategy) (bool, error) {
+	strategy, ok := s.(*v1alpha1.StrategyDetailsDeployment)
+	if !ok {
+		return false, fmt.Errorf("attempted to install %s strategy with deployment installer", strategy.GetStrategyName())
```
Comment (dtfranz) [inline pkg/controller/install/certresources.go:277]: Could we make this error a little more descriptive? To me it's just saying "we attempted an install" rather than "install failed because: ..." From what I can tell, looks like a failure due to a missing strategy? But I'm not super familiar with this code.

File: pkg/controller/install/certresources.go [hunk 4]
```diff
@@ -223,15 +230,74 @@ func (i *StrategyDeploymentInstaller) CertsRotated() bool {
...
+	}
+
+	hasCerts := map[string]struct{}{}
```
Comment (dtfranz) [inline pkg/controller/install/certresources.go:280]: nitpick and not required: we're using `k8s.io/apimachinery/pkg/util/sets` in v1 and inconsistently in v0. Might make things a little cleaner but it's not a big deal to me.

File: pkg/controller/install/certresources.go [hunk 5]
```diff
@@ -157,6 +158,13 @@ func ServiceName(deploymentName string) string {
 	return deploymentName + "-service"
 }
 
+func HostnamesForService(serviceName, namespace string) []string {
+	return []string{
+		fmt.Sprintf("%s.%s", serviceName, namespace),
+		fmt.Sprintf("%s.%s.svc", serviceName, namespace),
```
Comment (tmshort) [inline pkg/controller/install/certresources.go:164]: not adding `.cluster` nor `.cluster.local`?

File: pkg/controller/install/certresources.go [hunk 6]
```diff
@@ -223,15 +231,74 @@ func (i *StrategyDeploymentInstaller) CertsRotated() bool {
...
+		if hasCerts.Has(sddSpec.Name) {
+			certSecret, err := i.strategyClient.GetOpLister().CoreV1().SecretLister().Secrets(i.owner.GetNamespace()).Get(SecretName(ServiceName(sddSpec.Name)))
+			if err == nil {
+				if shouldRotateCerts(certSecret, HostnamesForService(ServiceName(sddSpec.Name), i.owner.GetNamespace())) {
+					return true, nil
+				}
+			} else if apierrors.IsNotFound(err) {
+				return true, nil
+			} else {
+				return false, err
```
Comment (tmshort) [inline pkg/controller/install/certresources.go:295]: If there's some other error, what happens? This does not rotate the certificate.

File: pkg/controller/operators/olm/operator.go [hunk 7]
```diff
@@ -2207,18 +2207,18 @@ func (a *Operator) transitionCSVState(in v1alpha1.ClusterServiceVersion) (out *v
 			return
 		}
 		if installErr := a.updateInstallStatus(out, installer, strategy, v1alpha1.CSVPhaseInstalling, v1alpha1.CSVReasonWaiting); installErr != nil {
-			// Re-sync if kube-apiserver was unavailable
-			if apierrors.IsServiceUnavailable(installErr) {
-				logger.WithError(installErr).Info("could not update install status")
-				syncError = installErr
-				return
-			}
```
Comment (tmshort) [inline pkg/controller/operators/olm/operator.go:2215]: Why was this moved lower? If there's an error, shouldn't we check it first?

**Review by tmshort — COMMENTED** [review]
Just a few questions right now.
However, _and I realize this is pre-existing code_, I don't like creating a certificate before checking it. The process should really be:
1. Does the cert exist? No -> create the cert
2. Is the cert need to be rotated? Yes -> rotate/create the cert

Right now, a cert is created _every time_ the rotate check is to be done, and that is a **BIG** waste of CPU.

File: pkg/controller/operators/olm/operator.go (hunk 7 above)
Comment (joelanford) [reply pkg/controller/operators/olm/operator.go:2215]: I had a similar thought. Maybe we shouldn't timeout the install plan if the reason for things taking so long is that KAS is unavailable.

File: pkg/controller/install/certresources.go [hunk 8]
```diff
@@ -223,15 +231,74 @@ func (i *StrategyDeploymentInstaller) CertsRotated() bool {
...
+	// Check validity of serving cert and if serving cert is trusted by the CA
+	for _, host := range hosts {
+		if err := certs.VerifyCert(ca, cert, host); err != nil {
+			return true
+		}
+	}
 	return false
 }
 
+func (i *StrategyDeploymentInstaller) ShouldRotateCerts(s Strategy) (bool, error) {
```
Comment (joelanford) [inline pkg/controller/install/certresources.go:275]: (Similar to @tmshort's comment about existing code and reluctance to change it)...

It looks like this function is setup as an all or nothing thing? If _any_ of the certs needs to be rotated, we're going to rotate all of them?

In practice, maybe this doesn't actually make a difference because all of the certs are created initially at the same time with the same expiration. Thus generally all or none will need rotating?

File: pkg/controller/install/certresources.go [hunk 9]
```diff
@@ -223,15 +231,74 @@ func (i *StrategyDeploymentInstaller) CertsRotated() bool {
...
+	strategy, ok := s.(*v1alpha1.StrategyDetailsDeployment)
+	if !ok {
+		return false, fmt.Errorf("failed to install %s strategy with deployment installer: unsupported deployment install strategy", strategy.GetStrategyName())
+	}
+
+	hasCerts := sets.NewString()
```
Comment (joelanford) [inline pkg/controller/install/certresources.go:281]: I believe this is deprecated. The `sets` package now supports/prefers the generic type.
```suggestion
	hasCerts := sets.New[string]()
```

File: pkg/controller/install/certresources.go (hunk 6 above)
Comment (joelanford) [reply pkg/controller/install/certresources.go:295]: I assume it means we enter exponential backoff (at least until we hit the 5m install plan timeout)?

Any other error is likely to be either temporary (in which case it's hopefully resolved next time around), or not (in which case we would likely fail in a similar way trying to rotate the cert).

The only scenario that is coming to mind where `Get` secret fails, but a subsequent `Create` or `Update` would succeed would be if we lack read permission but have write permission.

File: pkg/controller/install/certresources.go (hunk 8 above)
Comment (ankitathomas) [reply pkg/controller/install/certresources.go:275]: That's just the test to see if we need to transition the CSV from a Success/Failed state, there's a separate check before deciding to reuse/refresh certs for each deployment separately [here](https://github.com/operator-framework/operator-lifecycle-manager/pull/3217/files#diff-bc7b779225ba70ff0e1f63ff55cb2dd3d71b5ae4f96c158947266da9f8fde60dR381)

File: pkg/controller/install/certresources.go (hunk 5 above)
Comment (ankitathomas) [reply pkg/controller/install/certresources.go:164]: These should resolve even without the cluster domain specified, though we may need to test for the current cluster domain before adding `.cluster` or `.cluster.local` in case it's different from the default.

**Comment by ankitathomas** [comment]
>     * re-reconcile if the cert secret changes
We already do this for all [olm managed secrets](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/controller/operators/olm/operator.go#L430)

>     * re-reconcile at a time in the future base on the current secrets' expiration time.
With each reconcile, we check all certs for anything that expires in a day or less and rotate all of those, including the ones that are already expired. The problem was that we were checking the cert freshness timestamps on the CSV to make those checks, and those were sometimes being incorrectly updated when the cert rotate hadn't really succeeded.

File: pkg/controller/install/certresources.go (hunk 8 above)
Comment (joelanford) [reply pkg/controller/install/certresources.go:276]: Got it. Thanks!

**Comment by joelanford** [comment]
> With each reconcile, we check all certs for anything that expires in a day or less and rotate all of those, including the ones that are already expired.

But I'm wondering if there's a scenario where:
1. I install an operator that needs a cert
2. OLM creates the secrets, sets expiration to `now + N`
3. Everything reconciles, steady state is achieved, no further changes are made to the CSV.
4. `N` time passes, and the certs expire. Still no changes, so still no cert renewal?

Is there something that forces a re-reconcile inside the time window where:
- The certs are not yet expired, but
- They are close enough to expiration that we'll rotate them.

**Review by tmshort — APPROVED** [review]
/lgtm

Although we may want to look at the order of operations when a cert is to be rotated. As, this could lead to CPU spikes.
